File: flitetrakr_reactive/src/main/java/com/github/albertosh/flitetrakr/reactive/usecases/priceofconnection/PriceOfConnectionUseCase.java
package com.github.albertosh.flitetrakr.reactive.usecases.priceofconnection;

import com.google.common.base.Preconditions;

import com.github.albertosh.flitetrakr.model.Connection;
import com.github.albertosh.flitetrakr.reactive.services.ConnectionServiceError;
import com.github.albertosh.flitetrakr.reactive.services.IConnectionService;
import com.github.albertosh.flitetrakr.util.Pair;

import java.util.List;

import rx.Observable;
import rx.Single;
import rx.schedulers.Schedulers;

public class PriceOfConnectionUseCase
        implements IPriceOfConnectionUseCase {

    private final IConnectionService connectionService;

    public PriceOfConnectionUseCase(IConnectionService connectionService) {
        this.connectionService = Preconditions.checkNotNull(connectionService);
    }

    @Override
    public Single<Integer> execute(PriceOfConnectionUseCaseInput input) {
        final List<String> codes = input.getCodes();

        return Observable.create((Observable.OnSubscribe<Pair<String, String>>) subscriber -> {
            int connectionSize = codes.size();
            for (int i = 1; i < connectionSize; i++) {
                if (subscriber.isUnsubscribed())
                    return;
                Pair<String, String> cityPair = Pair.create(codes.get(i - 1), codes.get(i));
                subscriber.onNext(cityPair);
            }
            subscriber.onCompleted();
        })
                .flatMap(pair -> connectionService.recoverConnection(pair.first, pair.second)
                        .subscribeOn(Schedulers.io())
                        .toObservable())
                .map(Connection::getPrice)
                .reduce(0, (accum, price) -> accum + price)
                .onErrorResumeNext(error -> {
                    if (error instanceof ConnectionServiceError) {
                        ConnectionServiceError connectionServiceError = (ConnectionServiceError) error;
                        if (connectionServiceError.getErrorType().equals(ConnectionServiceError.ErrorType.CONNECTION_NOT_FOUND))
                            return Observable.error(PriceOfConnectionUseCaseError.connectionNotFound());
                    }
                    return Observable.error(error);
                }).toSingle();
    }


}

File: flitetrakr_sync_kotlin/src/main/kotlin/com/github/albertosh/flitetrakr/usecases/PriceOfConnectionUseCase.kt
package com.github.albertosh.flitetrakr.usecases

import com.github.albertosh.flitetrakr.services.IConnectionService
import com.github.albertosh.flitetrakr.util.language.LanguageUtils
import com.github.albertosh.flitetrakr.util.language.Message

interface IPriceOfConnectionUseCase {

    fun execute(input: PriceOfConnectionUseCaseInput): PriceOfConnectionUseCaseOutput

}

data class PriceOfConnectionUseCaseInput(val codes: List<String>)

data class PriceOfConnectionUseCaseOutput(val price: Int)

sealed class PriceOfConnectionUseCaseError(message : String) : RuntimeException(message){
    object connectionNotFound : PriceOfConnectionUseCaseError(
            LanguageUtils.getMessage(Message.CONNECTION_NOT_FOUND))
}


class PriceOfConnectionUseCase(
        private val service: IConnectionService)
: IPriceOfConnectionUseCase {

    override fun execute(input: PriceOfConnectionUseCaseInput): PriceOfConnectionUseCaseOutput {

        val codes = input.codes

        val price = codes
                .subList(1, codes.size)
                .mapIndexed { i, city -> Pair(codes[i], city) }
                .map {
                    service.recoverConnection(it.first, it.second)
                            ?: throw PriceOfConnectionUseCaseError.connectionNotFound
                }
                .map { it.price }
                .reduce { acc, price -> acc + price }

        return PriceOfConnectionUseCaseOutput(price)
    }

}